sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java

```diff
@@ -286,10 +289,13 @@ protected AvroCoder(Class<T> type, Schema schema) {
 
           @Override
           public DatumReader<T> initialValue() {
-            return myCoder.getType().equals(GenericRecord.class)
-                ? new GenericDatumReader<>(myCoder.getSchema())
-                : new ReflectDatumReader<>(
-                    myCoder.getSchema(), myCoder.getSchema(), myCoder.reflectData.get());
+            if (myCoder.getType().equals(GenericRecord.class)) {
+              return new GenericDatumReader<>(myCoder.getSchema());
+            } else if (SpecificRecord.class.isAssignableFrom(myCoder.getType())) {
```

Comment:
This is fun: if the AvroCoder is instantiated as `AvroCoder<MySpecificRecord>` it will behave differently than `AvroCoder<GenericRecord>`, even though MySpecificRecord is a GenericRecord.  I can see this being a gotcha for some developers, but I think it's the right thing to do and I don't see an alternative.